fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/JobsServlet.java

```diff
@@ -431,10 +440,85 @@ protected void handleResultsResponse(
             );
             return Observable.error(responseException);
         }
+
+        if (jobsApiRequest.getPaginationParameters().isPresent()) {
+            Pagination<Result> pages = new AllPagesPagination<>(
+                    preResponse.getResultSet(),
+                    jobsApiRequest.getPaginationParameters().get()
+            );
+            responseContext = paginateResponseContext(responseContext, jobsApiRequest, pages);
+            PreResponse paginatedPreResponse = new PreResponse(
+                    new ResultSet(pages.getPageOfData(), preResponse.getResultSet().getSchema()),
+                    responseContext
+            );
+            return Observable.just(paginatedPreResponse);
+        }
         return Observable.just(preResponse);
     }
 
     /**
+     * Add pagination details to ResponseContext.
+     *
+     * @param responseContext  ResponseContext object contains all the meta info of the resultSet
+     * @param jobsApiRequest  JobsApiRequest object with all the associated info with it
+     * @param pages  Paginated resultSet
+     *
+     * @return Updated ResponseContext contains pagination info
+     */
+    protected ResponseContext paginateResponseContext(
+            ResponseContext responseContext,
+            ApiRequest jobsApiRequest,
+            Pagination<Result> pages
+    ) {
+        LinkedHashMap<String, URI> bodyLinks = new LinkedHashMap<>();
+        Arrays.stream(PaginationLink.values())
+                .forEachOrdered(link -> addLink(
+                                link,
+                                pages,
+                                jobsApiRequest.getUriInfo().getRequestUriBuilder(),
+                                bodyLinks)
+                );
```

Comment:
And if we're just copy-pasting this code, we shouldn't be... we should abstract it out into something reusable, otherwise we'll update one but forget the other and then feel bug pain down the road.